### ISSUE TITLE
When the model training is completed, a method is needed to test the training results.

### ISSUE THREAD

**Issue by llxxxll** [sql-machine-learning/sqlflow#390]
The solution I want:
Add a testing action or function in training SQL
```sql
SELECT *
FROM iris.train
TRAIN DNNClassifier

...
TESTING iris.test
...

LABEL class
INTO sqlflow_models.my_dnn_model;

```

Log addition:
```
Training set accuracy: 0.96721
Testing  set accuracy: 0.90132
Done training
```

**Comment by tonyyang-svail** [comment]
I think you mean the validation data instead of the testing data. I've never seen people include the testing data in a training process.

One possible solution is splitting the train table into training data (say 80% of the rows) and validation data (20%). And report the accuracy as
```
Training set accuracy: 0.96721
Validation set accuracy: 0.90132
Done training
```
So the user still writes
```
SELECT *
FROM iris.train
TRAIN DNNClassifier
LABEL class
INTO sqlflow_models.my_dnn_model;
```

**Comment by llxxxll** [comment]
How to define the data table needed for verification accuracy, Does SQLFlow automatically divide the training set and test set?

**Comment by tonyyang-svail** [comment]
I would suggest splitting training and validation automatically. Splitting is a common practice in ML and should be wrapped inside SQLFlow.

If a user needs to verify using a separate table, he can do `PREDICT` and save the result in a temporary table. Then compare the temporary table with ground truth table with a `select count(*) from t1, t2 where t1.id = t2.id and t1.class = t2.class` statement.

**Comment by llxxxll** [comment]
Great, this solved my problem.

**Comment by typhoonzero** [comment]
> I would suggest splitting training and validation automatically. Splitting is a common practice in ML and should be wrapped inside SQLFlow.

I'm afraid that auto splitting may encounter difficulties in some cases like:

To predict CTR, we may need to train using data in the last hour and validate using data in the comming 5 minutes.

For image classification, we need a validation set contains all classes.

The splitting strategy may differ when the job differs. Or we can just leave it to users to split the dataset using SQL.

**Comment by tonyyang-svail** [comment]
@typhoonzero I agree that auto splitting may not work well for all cases. However, I would argue automation is the core value of SQLFlow. Let's start with implementing the simplest splitting strategy, and then gradually move on to handle more complicated cases.

After discussing with @uuleon, we may implement the random splitting in the following way.

Say SQLFlow are dealing with the following SQL to train an ML model.
```SQL
SELECT col1, col2, col3
FROM mytable
TRAN ...
```

The data comes from the standard select part `SELECT col1, col2, col3 FROM mytable`, and let's say the query result looks like the following
<table>
  <tr>
    <td>col1</td>
    <td>col2</td>
    <td>col3</td>
  </tr>
  <tr>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
  </tr>
  <tr>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
  </tr>
  <tr>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
  </tr>
  <tr>
    <td>...</td>
    <td></td>
    <td></td>
  </tr>
</table>

We want to split the result into 80% training data and 20% validation data.

**First, we add a column `train_val_split` via `RAND()` and save the result to a temporary table.** Note the `RAND()` function returns a random number between 0 (inclusive) and 1. The result temporary table looks like the following.

<table>
  <tr>
    <th colspan="4">temp_table</th>
  </tr>
  <tr>
    <td>col1</td>
    <td>col2</td>
    <td>col3</td>
    <td>train_val_split</td>
  </tr>
  <tr>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>0.3</td>
  </tr>
  <tr>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>0.9</td>
  </tr>
  <tr>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>0.5</td>
  </tr>
  <tr>
    <td>...</td>
    <td></td>
    <td></td>
    <td></td>
  </tr>
</table>

We can generate the corresponding SQL using the following code template
```SQL
CREATE TABLE {.TempTableName} AS (
    SELECT *, RAND() AS train_val_split FROM (
        {.StandardSQL}
    )
)
```

**Second, we fetch the training/validation data using two different queries respectively.** The query for training data can be written as `SELECT * FROM temp_table WHERE train_val_split <= 0.8`, which fetches row1 and row3 etc.. The query for validation data can be written as `SELECT * FROM temp_table WHERE train_val_split > 0.8`, which fetches the rest of the rows.

With these two SQL statements, we can create the training dataset and validation dataset accordingly. At the end of the training job, we can clean up the temporary table.
